### PR TITLE
Feature/expand data collection from prs

### PR DESCRIPTION
**Pull request by xtuchyna**
## Related Issues and Dependencies

…

## This introduces a breaking change

- [x] Yes
- [ ] No

This PR introduces a different way of saving knowledge files for repos.
For every project a new directory is created, this is the same for every repo in each project.

For each analysed repo files issues, pull-requests and pull_requests and pull_request_reviews should be created.

## This Pull Request implements
Analysis of repository in means of issues (that are not PRs), PRs, and PR reviews.

### CODE REVIEW

File: create_bot_knowledge.py [hunk 1]
```diff
@@ -80,139 +82,243 @@ def connect_to_source(project: Tuple[str, str]):
...
+        #"time_to_approve": time_to_approve,
+        "closed_at": closed_at,
+        "closed_by": pull.as_issue().closed_by.login,
+        "time_to_close": closed_at - created_at,
+        "merged_at": merged_at,
+        "commits_number": commits, 
+https://trello.com/c/VRyizGlU/846-data-collection-from-github-repositories-nokr
+https://trello.com/c/VRyizGlU/846-data-collection-from-github-repositories-nokrts()),
+https://trello.com/c/VRyizGlU/846-data-collection-from-github-repositories-nokr
+https://trello.com/c/VRyizGlU/846-data-collection-from-github-repositories-nokr
```
Comment (pacospace) [inline create_bot_knowledge.py]: Not sure what is this :)

File: create_bot_knowledge.py [hunk 2]
```diff
@@ -21,20 +21,22 @@
...
+from github import Github, GithubObject, Issue, IssueComment, PullRequest, PullRequestReview
+from github.Repository import Repository
 
 _LOGGER = logging.getLogger(__name__)
 logging.basicConfig(level=logging.INFO)
 
 _GITHUB_ACCESS_TOKEN = os.getenv("GITHUB_ACCESS_TOKEN")
 
-PROJECTS = [
+PROJECTS = os.getenv("PROJECTS") #[
```
Comment (pacospace) [inline create_bot_knowledge.py]: I like the idea, but I don't see where you actually use `PROJECTS`

File: create_bot_knowledge.py (hunk 1 above)
Comment (xtuchyna) [reply create_bot_knowledge.py]: what the hell is this :D embarrassing, will remove this

File: create_bot_knowledge.py (hunk 2 above)
Comment (xtuchyna) [reply create_bot_knowledge.py]: i've changed main to the function which takes projects as parameter, will remove this env variable

File: create_bot_knowledge.py [hunk 3]
```diff
@@ -21,24 +21,277 @@
...
+                  'resolve',
+                  'resolves',
+                  'resolved'}
+
+
+standalone_labels = {'size'}
+
+
+def get_labeled_size(labels: List[str]) -> str:
+    for label in labels:
```
Comment (pacospace) [inline create_bot_knowledge.py]: There are missing descriptions, also in the functions below.

File: create_bot_knowledge.py [hunk 4]
```diff
@@ -21,24 +21,277 @@
...
+        "time_to_close": closed_at - created_at,
+        "merged_at": merged_at,
+        "commits_number": commits,
+        "referenced_issues": get_referenced_issues(pull),
+        "interactions": get_interactions(pull.get_comments()),
+    }
+
+
+def pull_request_analysis(project: Repository, project_knowledge: Path):
+    """Analysis of every closed pullrequest in repository
```
Comment (pacospace) [inline create_bot_knowledge.py]: Why there are two functions for PR analysis?

File: create_bot_knowledge.py [hunk 5]
```diff
@@ -21,24 +21,277 @@
...
+
+def check_directory(knowledge_dir: Path):
+    """Check if directory for bot knowledge exists. If not, create one."""
+    if not knowledge_dir.exists():
+        _LOGGER.info(
+            "No knowledge from any repo has ever been created, creating new directory at %s" % knowledge_dir)
+        os.makedirs(knowledge_dir)
+
+
+issue_keywords = {'close',
```
Comment (pacospace) [inline create_bot_knowledge.py]: This variable and other below can be named in capital letters and moved to the beginning of the file.

Comment (xtuchyna) [reply create_bot_knowledge.py]: isn't it better to have them at the place we actually use them? I know that the constants tend to be at the beginning, but this seems like a more readable option to me.

Comment (pacospace) [reply create_bot_knowledge.py]: We have standards for a reason :)

Comment (xtuchyna) [reply create_bot_knowledge.py]: yeah i meant that indeed it should be named as constant (capitals + underscores)
but should it be at the beginning of a file? there seems to be no standard on that, or is it?
https://www.python.org/dev/peps/pep-0008/#constants

Comment (pacospace) [reply create_bot_knowledge.py]: You might need to reuse them in another file, it's easier to have them all at the beginning.

Comment (fridex) [reply create_bot_knowledge.py]: It's a convention (I don't recall any standard on this). BTW, if you do not plan to adjust the set, you can declare it as a frozenset, as you use typing, you can declare it as final - something like (I haven't tried to run the code):

```python
from typing import Final
from typing import FrozenSet

ISSUE_KEYWORDS: Final[FrozenSet[str]] = frozenset({
                  'close',
                  'closes',
                  'closed',
                  'fix',
                  'fixes',
                  'fixed',
                  'resolve',
                  'resolves',
                  'resolved'
})

**Comment by pacospace** [comment]
recheck

File: create_bot_knowledge.py [hunk 6]
```diff
@@ -21,174 +21,379 @@
...
+        return
+
+    for idx, issue in enumerate(new_issues, 1):
+        _LOGGER.info("Analysing ISSUE no. %d/%d" % (idx, len(new_issues)))
+        store_issue(issue, data)
+
+    save_knowledge(data_path, data)
+
+
+def extract_pullrequest_review_requests(pullrequest: PullRequest) -> List[str]:
```
Comment (pacospace) [inline create_bot_knowledge.py]: what about `pull_request`? :) Also in other places.

File: create_bot_knowledge.py [hunk 7]
```diff
@@ -21,174 +21,379 @@
...
+
+STANDALONE_LABELS = {'size'}
+
+
+def connect_to_source(project: Tuple[str, str]) -> Repository:
     """Connect to GitHub.
 
     :param project: Tuple source repo and repo name.
     """
     # TODO: It should use only one library for source.
```
Comment (pacospace) [inline create_bot_knowledge.py]: if you are using only one library, we can remove also this  comment :)

**Pull request merged**